### PR TITLE
feature: Add RPC Selector

### PR DESCRIPTION
**Pull request by ramos-ph**

### CODE REVIEW

File: test/services/RPCSelector.test.ts [hunk 1]
```diff
@@ -19,25 +19,46 @@ describe('RPCSelector', () => {
...
+      describe('when the preferred URL is not up to date', () => {
+        it.skip('returns the RPC with highest block number', async () => {
+          const urls = [
+            'https://data-seed-prebsc-1-s1.binance.org:8545/',
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+          ];
+          const rpcSelector = new RPCSelector(urls);
+          expect(await rpcSelector.apply()).to.eq(urls[1]);
+        });
+      });
```
Comment (dougmrqs) [inline test/services/RPCSelector.test.ts:50]: I'd like to see the block number mocked to see if it actually selected the highest block number.

And also, what happens if they both return the same block number?

File: test/services/RPCSelector.test.ts [hunk 2]
```diff
@@ -19,25 +19,46 @@ describe('RPCSelector', () => {
...
+      // Skipping because the RPCs results are not always reliable
+      describe('when the preferred URL is up to date', () => {
+        it.skip('returns the preferred URL', async () => {
+          const urls = [
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+            'https://data-seed-prebsc-2-s3.binance.org:8545/',
+          ];
+          const rpcSelector = new RPCSelector(urls);
+          expect(await rpcSelector.apply()).to.eq(urls[0]);
+        });
```
Comment (dougmrqs) [inline test/services/RPCSelector.test.ts:38]: Can you make the test reliable?

File: test/services/RPCSelector.test.ts [hunk 3]
```diff
@@ -19,25 +19,46 @@ describe('RPCSelector', () => {
...
+      describe('when RPC exceeds request time threshold', () => {
+        it('timeouts', async () => {
+          const urls = [
+            'https://data-seed-prebsc-1-s1.binance.org:8545/',
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+          ];
+
+          const rpcSelector = new RPCSelector(urls, 1);
+          expect(await rpcSelector.apply()).to.eq(urls[1]);
+        });
```
[the remark this answers is not in the document]
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts:61]: Yes

File: test/services/RPCSelector.test.ts (hunk 1 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts:50]: > And also, what happens if they both return the same block number?

The first one is selected. Default `reduce` behavior

File: test/services/RPCSelector.test.ts (hunk 2 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts:38]: Yes

**Review by matheushahnn — APPROVED** [review]
looks good to me, just addressed the Douglas comments

File: test/services/RPCSelector.test.ts [hunk 4]
```diff
@@ -19,25 +21,111 @@ describe('RPCSelector', () => {
...
-        const rpcSelector = new RPCSelector(urls);
-        expect(await rpcSelector.apply()).to.eq(urls[0]);
+    describe('when there are multiple URLs', () => {
+      describe('when the preferred URL is up to date', () => {
+        const stubValue = {
+          timestamp: Math.floor(Date.now() / 1000) - 30,
+        } as Block;
+
+        it('returns the preferred URL', async () => {
+          const mockedProvider = sinon.stub(JsonRpcProvider.prototype, 'getBlock').resolves(stubValue);
```
Comment (dougmrqs) [inline test/services/RPCSelector.test.ts:37]: this mock mocks for both URL calls, right?

File: test/services/RPCSelector.test.ts [hunk 5]
```diff
@@ -19,25 +21,111 @@ describe('RPCSelector', () => {
...
+        it('returns the RPC with highest block number', async () => {
+          const mockedGetBlock = sinon.stub(JsonRpcProvider.prototype, 'getBlock').resolves(stubValue);
+          const mockedGetBlockNumber = sinon.stub(JsonRpcProvider.prototype, 'getBlockNumber').resolves(100_000);
+
+          const urls = [
+            'https://data-seed-prebsc-1-s1.binance.org:8545/',
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+          ];
+          const rpcSelector = new RPCSelector(urls, 100);
+          expect(await rpcSelector.apply()).to.eq(urls[1]);
```
Comment (dougmrqs) [inline test/services/RPCSelector.test.ts:64]: nl before expect

File: test/services/RPCSelector.test.ts (hunk 4 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts:37]: It only needs to mock one. The other one will not execute when the timestamp is up-to-date.

File: test/services/RPCSelector.test.ts [hunk 6]
```diff
@@ -19,25 +21,111 @@ describe('RPCSelector', () => {
...
+          });
+        };
+
+        const fakeGetBlockNumber = (): Promise<number> => {
+          return new Promise((resolve) => {
+            setTimeout(resolve, 5000, 100_000);
+          });
+        };
+
+        it('timeouts and select the first of the non-preferred RPCs', async () => {
```
Comment (dougmrqs) [inline test/services/RPCSelector.test.ts:85]: 'timeouts and selects....'

File: test/services/RPCSelector.test.ts (hunk 4 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts:37]: We call `getBlock` on the first one and `getBlockNumber` on the rest of the URLs.

File: test/services/RPCSelector.test.ts [hunk 7]
```diff
@@ -19,25 +21,111 @@ describe('RPCSelector', () => {
...
+            'https://data-seed-prebsc-1-s1.binance.org:8545/',
+            'https://data-seed-prebsc-1-s2.binance.org:8545/',
+          ];
+
+          const rpcSelector = new RPCSelector(urls, 500);
+          expect(await rpcSelector.apply()).to.eq(urls[1]);
+
+          mockedGetBlock.restore();
+          mockedGetBlockNumber.restore();
+        });
```
Comment (dougmrqs) [inline test/services/RPCSelector.test.ts:101]: which of the two timed out?

File: test/utils/helpers.test.ts [hunk 8]
```diff
@@ -20,3 +18,25 @@ it('evenHex', () => {
...
 });
+
+describe('isTimestampMoreRecentThan', () => {
+  const timestamp = Math.floor(Date.now() / 1000);
+
+  describe('when timestamp is greater the current time', () => {
+    it('returns false', () => {
+      expect(isTimestampMoreRecentThan(timestamp + 1000, 60)).to.be.false;
+    });
+  });
```
Comment (dougmrqs) [inline test/utils/helpers.test.ts:29]: ... greater than ...

We set timestamp + 1000 to see if it is more recent than 60 second. Are we looking 1000 seconds in the future and it is not more recent than 60 seconds? I don't get it.

File: test/utils/helpers.test.ts [hunk 9]
```diff
@@ -20,3 +18,25 @@ it('evenHex', () => {
...
+
+  describe('when timestamp is not between the current date and the the current date minus the limit', () => {
```
Comment (dougmrqs) [inline test/utils/helpers.test.ts:31]: typo at the description, double `the`

File: test/utils/helpers.test.ts [hunk 10]
```diff
@@ -20,3 +18,25 @@ it('evenHex', () => {
...
+    it('returns false', () => {
+      expect(isTimestampMoreRecentThan(timestamp - 120, 60)).to.be.false;
+    });
+  });
+
+  describe('when timestamp is between the current date and the current date minus the limit', () => {
```
Comment (dougmrqs) [inline test/utils/helpers.test.ts:37]: what is the limit? 60?

File: test/services/RPCSelector.test.ts (hunk 7 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts:101]: It timeouts both. The reduce will take care of selecting the first one of the non-preferred URLs.

File: test/utils/helpers.test.ts (hunk 10 above)
Comment (ramos-ph) [reply test/utils/helpers.test.ts:37]: Yes. It's the second argument.

File: test/utils/helpers.test.ts (hunk 8 above)
Comment (ramos-ph) [reply test/utils/helpers.test.ts:29]: Looking at the implementation, the timestamp cannot be greater than the current date.

File: test/utils/helpers.test.ts (hunk 9 above)
Comment (ramos-ph) [reply test/utils/helpers.test.ts]: Thanks

File: test/services/RPCSelector.test.ts (hunk 6 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts]: Ty

File: test/services/RPCSelector.test.ts (hunk 5 above)
Comment (ramos-ph) [reply test/services/RPCSelector.test.ts]: Fixed

**Review by dougmrqs — APPROVED** [review]
LGTM.

File: package.json [hunk 11]
```diff
@@ -35,6 +35,7 @@
     "@types/mocha": "~8.2.0",
     "@types/moxios": "~0.4.9",
     "@types/node": "~14.14.13",
+    "@types/sinon": "^10.0.8",
```
Comment (DZariusz) [inline package.json:38]: use ~ everywhere

File: src/services/RPCSelector.ts [hunk 12]
```diff
@@ -31,33 +29,37 @@ class RPCSelector {
...
     } catch (e) {
       return false;
     }
   }
 
   private isBlockRecentlyMinted(block: Block): boolean {
-    const currentDateInSeconds = Math.floor(Date.now() / 1000),
-      oneMinute = 60;
-    return block.timestamp - (currentDateInSeconds - oneMinute) <= oneMinute;
+    return isTimestampMoreRecentThan(block.timestamp, 60);
```
Comment (DZariusz) [inline src/services/RPCSelector.ts:40]: let's make 60 as param for constructor

Comment (ramos-ph) [reply src/services/RPCSelector.ts:40]: What should we call it?

**Pull request merged**